templates/etcd/etcd.py

```diff
@@ -0,0 +1,113 @@
+
+    def __init__(self, name=None, guid=None, data=None):
+        super().__init__(name=name, guid=guid, data=data)
+        self.add_delete_callback(self.uninstall)
+        self.recurring_action('_monitor', 30)  # every 30 seconds
+
+    def validate(self):
+        self.state.delete('status', 'running')
+        for nic in self.data['nics']:
+            if nic['type'] == 'zerotier':
+                break
+        else:
+            raise ValueError('Service must contain at least one zerotier nic')
+
+    def _deploy(self):
+        etcd_sal = self._etcd_sal
+        etcd_sal.deploy()
+        self.data['ztIdentity'] = etcd_sal.zt_identity
+        
+
+    def _monitor(self):
+        self.logger.info('Monitor etcd %s' % self.name)
+        self.state.check('actions', 'start', 'ok')
+
+        if not self._etcd_sal.is_running():
+            self.state.delete('status', 'running')
+            self._deploy()
+            self._etcd_sal.start()
+            if self._etcd_sal.is_running():
+                self.state.set('status', 'running', 'ok')
+        else:
+            self.state.set('status', 'running', 'ok')
+
+    @property
+    def _node_sal(self):
+        return j.clients.zos.get(NODE_CLIENT)
+
+    @property
+    def _etcd_sal(self):
+        kwargs = {
+            'name': self.guid,
+            'node': self._node_sal,
+            'zt_identity': self.data['ztIdentity'],
+            'nics': self.data['nics'],
+            'token': self.data['token'],
+            'cluster': self.data['cluster'],
+        }
+        return j.sal_zos.etcd.get(**kwargs)
+
+
+    def install(self):
+        self.logger.info('Installing etcd %s' % self.name)
+        self._deploy()
+        self.state.set('actions', 'install', 'ok')
+
+    def start(self):
+        """
+        start etcd server
+        """
+        self.state.check('actions', 'install', 'ok')
+        self.logger.info('Starting etcd %s' % self.name)
+        self._deploy()
+        self._etcd_sal.start()
+        self.state.set('actions', 'start', 'ok')
+        self.state.set('status', 'running', 'ok')
+
+    def stop(self):
+        """
+        stop etcd server
+        """
+        self.state.check('actions', 'install', 'ok')
+        self.logger.info('Stopping etcd %s' % self.name)
+        self._etcd_sal.stop()
+        self.state.delete('actions', 'start')
+        self.state.delete('status', 'running')
+
+    def uninstall(self):
+        self.logger.info('Uninstalling etcd %s' % self.name)
+        self._etcd_sal.destroy()
+        self.state.delete('actions', 'install')
+        self.state.delete('status', 'running')
+
+    def connection_info(self):
+        self.state.check('status', 'running', 'ok')
+        etcd_sal = self._etcd_sal
+        return {'peer_url': etcd_sal.peer_url, 'client_url': etcd_sal.client_url}
+    
+    def update_cluster(self, cluster):
+        self.data['cluster'] = cluster
+    
+    def insert_record(self, key, value):
```

Comment:
these 3 methods doesn't belong here. I someone need to put record in etcd it needs to connect to it using a client

Reply:
I disagree. Wouldnt that mean duplicate code every where were we want to insert anything into etcd? what's the point of the services and sal if other services will connect using the client? Both coredns and traefik insert data into etcd and I don't see any reason why it shouldnt be done through the service.

Reply:
You are mixing stuff. A service is mean to del with the lifecycle of a application. It is not meant to provide an interface to talk to the application itself. 
And by having this here, we ARE duplicating code. While if you use the client, there is only one interface to talk to etcd, and it is using the client.

Reply:
Ok so I've got more information about all this. It is another component that will connect to etcd and handle the management of the domain and virtual host. So there is no need to have such method on the services themself